### PR TITLE
feat(client-debugger): Add Chrome devtools extension prototype

### PR DESCRIPTION
**Pull request by Josmithr**
Updates chrome-extension library to be a devtools extension. Also updates the core debugger package to communicate Container summary details via message passing (+ related infrastructure), and updates visuals library to make some components reusable across direct-Container (i.e. the simple React library model) and extension scenarios.

Future work will add message-passing capabilities for other Container state / data.

### Notes for reviewers:

This PR touched a lot of files, but the primary changes here are:
1. The overhaul of the `client-debugger-chrome-extension` package to include multiple extension prototype variants for consideration (see the package README for details on the variants).
2. The message-passing-based eventing added to `Registry.ts` and `FluidClientDebugger.ts` (+ related infrastructure)

#### Other notes

1. The Chrome extension path is pretty orthogonal to the existing React library path we took for the early prototype. Both views want to visualize (mostly) the same data, but they have to access it in different ways. I have attempted to reuse components from the existing `client-debugger-view` package in the chrome extension logic as much as was reasonable, but there is still some code that will appear to be duplicated between the packages. Once we have a definitive pathway for our MVP, we can likely begin removing code from the `view` package (or possibly even remove that package altogether, and move the needed pieces into the extension package).

3. I tried to document the message passing contracts on `Registry` and `FluidClientDebugger`, but since none of the message-passing aspects appear on the API surface, I don't know how useful / discoverable they are. Definitely interested in feedback / suggestions related to surfacing this better.
    - Suggestions will likely become follow-up work items to be addressed in the future, since this is still an early prototype, but I am definitely interested in any feedback folks have now :)

### CODE REVIEW

File: packages/tools/client-debugger/client-debugger-view/src/test/app/App.tsx [hunk 1]
```diff
@@ -211,8 +213,11 @@ export function App(): React.ReactElement {
 	const { privateContainer, sharedContainer } = useContainerInfo();
 
 	const view = (
-		<Stack horizontal>
-			<StackItem>
+		<Stack horizontal style={{ height: "100%" }}>
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/test/app/App.tsx:216]: TODO: Revert these changes before merging

File: packages/tools/client-debugger/client-debugger-view/src/RenderClientDebugger.ts [hunk 2]
```diff
@@ -9,38 +9,23 @@ import { FluidClientDebuggers } from "./Debugger";
 
 /**
  * Renders Fluid client debug view by appending it to the provided DOM element.
- * Will fast return false if the input is `null`.
  *
  * @param targetElement - The HTML element takes the client debugger view.
  *
- * @returns `true` if the debug view was succesfully rendered, otherwise `false`.
+ * @returns `true` if the debug view was successfully rendered, otherwise `false`.
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/RenderClientDebugger.ts:15]:
```suggestion
 * @returns A promise that resolves once the debugger view has been rendered for the first time.
```

File: packages/tools/client-debugger/client-debugger-chrome-extension/src/content-extension/ContentScript.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+function hide(): void {
+	// TODO: suspend message subscription for the debugger itself
+
+	document.querySelector(`#${debuggerPanelId}`)?.remove();
+}
+
+// #region Background <- -> Content script messaging
+
+// TODO: differentiate senders (registry / debugger or background script)
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-chrome-extension/src/content-extension/ContentScript.ts:43]:
```suggestion
```

File: packages/tools/client-debugger/client-debugger-chrome-extension/src/devtools-extension/messaging/Utilities.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	message: TMessage,
+	messageSource: string,
+	targetPort: TypedPortConnection<TMessage>,
+	loggingOptions?: MessageLoggingOptions,
+): void {
+	console.log(
+		formatMessageForLogging(
+			`Relaying "${message.type}" message from "${messageSource}" to port "${targetPort.name}":`,
+			loggingOptions,
+		),
```
[the remark this answers is not in the document]
Comment (Josmithr) [reply packages/tools/client-debugger/client-debugger-chrome-extension/src/devtools-extension/messaging/Utilities.ts]: Resolved.

File: packages/tools/client-debugger/client-debugger-chrome-extension/src/content-extension/messaging/WindowConnection.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	 * Messages are forwarded on to subscribers for valid {@link IDebuggerMessage}s from the expected source.
+	 */
+	private onWindowMessageEvent(event: MessageEvent<Partial<IDebuggerMessage>>): void {
+		const message = event.data;
+
+		// Only relay message if it is one of ours, and if the source is the window's debugger
+		// (and not a message originating from the extension).
+		if (isDebuggerMessage(message) && message.source === debuggerMessageSource) {
+			console.log(
+				formatForLogging(`Relaying "${message.type}" message from Window to Extension:`),
```
[the remark this answers is not in the document]
Comment (Josmithr) [reply packages/tools/client-debugger/client-debugger-chrome-extension/src/content-extension/messaging/WindowConnection.ts]: Resolved

**Comment by alexvy86** [comment]
I'm going through it either way, but my first piece of feedback is: how can we split this into multiple smaller PRs? 😄

**Comment by Josmithr** [comment]
> I'm going through it either way, but my first piece of feedback is: how can we split this into multiple smaller PRs? 😄

Fair question. I could probably split out the changes I made to the `view` package. I think the changes to the core debugger package and the extension package changes are pretty tied together though. In theory I could PR the core debugger message-passing changes separately, but the extension changes are what offer context there.

File: packages/tools/client-debugger/client-debugger-chrome-extension/src/messaging/Utilities.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+/**
+ * Relays the provided message to the specified target port.
+ *
+ * @remarks Thin wrapper to provide some message-wise type-safety, and to inject some automated logging.
+ *
+ * @internal
+ */
+export function relayMessageToRuntime<TMessage extends IDebuggerMessage>(
```
Comment (alexvy86) [inline packages/tools/client-debugger/client-debugger-chrome-extension/src/messaging/Utilities.ts:90]: Apparently unused? Same with `postMessageToRuntime` below.

**Review by alexvy86 — APPROVED** [review]
Overall I don't have any feedback to block this. Most of the comments are small things and can even be addressed in follow-up PRs.

File: packages/tools/client-debugger/client-debugger-chrome-extension/src/background/Logging.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+ * Background Script context label for logging.
+ */
+export const backgroundScriptLoggingContext = "BACKGROUND";
+
+/**
+ * Configuration for console logging.
+ */
+export const backgroundScriptMessageLoggingOptions: MessageLoggingOptions = {
+	context: backgroundScriptLoggingContext,
+};
```
[the remark this answers is not in the document]
Comment (Josmithr) [reply packages/tools/client-debugger/client-debugger-chrome-extension/src/background/Logging.ts:18]: I agree. I definitely want to add a more formal mechanism for console logging to the library.

File: packages/tools/client-debugger/client-debugger/src/ContainerMetadata.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * Metadata describing a {@link @fluidframework/container-definitions#IContainer}'s core state.
+ *
+ * @public
+ */
+export interface ContainerStateMetadata extends ContainerMetadata {
+	/**
+	 * Whether or not the Container has been closed (disposed).
+	 */
+	closed: boolean;
```
[the remark this answers is not in the document]
Comment (Josmithr) [reply packages/tools/client-debugger/client-debugger/src/ContainerMetadata.ts:43]: Yeah, I noticed that recently. I believe @chentong7 is going to be updating this as a part of her container state change visualization work.

File: packages/tools/client-debugger/client-debugger-chrome-extension/src/messaging/Utilities.ts (hunk 6 above)
Comment (Josmithr) [reply packages/tools/client-debugger/client-debugger-chrome-extension/src/messaging/Utilities.ts:90]: Good catch! I forgot about these. Will remove.

File: packages/tools/client-debugger/client-debugger/src/Registry.ts [hunk 9]
```diff
@@ -69,13 +88,86 @@ export interface DebuggerRegistryEvents extends IEvent {
...
 	public constructor() {
 		super();
+
+		// Register listener for inbound messages from the window (globalThis)
+		globalThis.addEventListener?.("message", this.windowMessageHandler);
+
+		// Initiate message posting of registry updates.
+		// TODO: Only do this after some external request?
+		this.on("debuggerClosed", this.postRegistryChange);
+		this.on("debuggerClosed", this.postRegistryChange);
```
[the remark this answers is not in the document]
Comment (Josmithr) [reply packages/tools/client-debugger/client-debugger/src/Registry.ts]: Copy-paste strikes again. This should have been binding to 2 different events. Fixed.

File: packages/tools/client-debugger/client-debugger-chrome-extension/README.md [hunk 10]
```diff
@@ -33,3 +44,6 @@ Use of Microsoft trademarks or logos in modified versions of this project must n
 <!-- AUTO-GENERATED-CONTENT:END -->
 
 <!-- Links -->
+
+[@fluid-tools/client-debugger]: https://github.com/microsoft/FluidFramework/tree/main/packages/tools/client-debugger/client-debugger
+[@fluidframework/client-debugger-view]: https://github.com/microsoft/FluidFramework/tree/main/packages/tools/client-debugger/client-debugger-view
```
[the remark this answers is not in the document]
Comment (Josmithr) [reply packages/tools/client-debugger/client-debugger-chrome-extension/README.md:48]: I did a pass through the README to clean things up / better explain some things. They are referenced now post-changes.

File: packages/tools/client-debugger/client-debugger/src/messaging/Messages.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+/**
+ * Message structure expected for window event listeners used by the Fluid Client Debugger.
+ *
+ * @public
+ */
+export interface IDebuggerMessage<TData = unknown> {
+	/**
+	 * The source of the event.
```
[the remark this answers is not in the document]
Comment (Josmithr) [reply packages/tools/client-debugger/client-debugger/src/messaging/Messages.ts]: I added a remarks block with some extra details, but I would like to avoid documenting anything here explicitly in terms of the devtools extension, since that is just one consumer.
